### DISCUSSION TITLE
Welcome to awesome-structure-editors Discussions!

### DISCUSSION THREAD

**Discussion by yairchu** [yairchu/awesome-structure-editors#24] [Announcements]
## 👋 Welcome!
We’re using Discussions as a place to connect with folks interested in structure editors!

We hope that you:
* Discuss structural editing and visions for a bright future of programming 🦄
* Share your experiences using structural editors
* Engage with other community members.
* Talk about your projects and hopefully find contributors
* Welcome others and are open-minded. Remember that this is a community we build together 💪.

To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by michaelsippel** [comment]
Hey, whatup!

This discussion was recommended by github and since I'm quite interested in structure editors, I thought this might me a good opportunity to join.
Thanks a lot for your efforts to encourage discussion & community in this space and to ease research with your project-list! Very much appreciated!

Over the course of roughly the last decade I worked on various software projects, mainly OS, Web and HPC, and recently a lot of 3d-modeling with OpenScad.   From that experiences -- collected frustrations I should say - a lot of ideas and fantasies came about how to improve programming in general and gradually lead me to the idea of syntax-based editing and eventually starting my own syntax-based-editor-project.

My notion is that, the big difference to traditional editing is, that human-machine interaction should be *keyboard-driven*, not *text-driven*.

Currently, I'm using the structure-editor library I wrote in rust to create a shell-like command interpreter where you enter the commands via syntax-based editing and pipelines are type checked with a novel type-system that is able to capture the multi-layeredness of encodings through so called *ladder-types* which induce a structural subtyping relation.
I believe that such a type system will be useful not only in a shell-scripting environment but also for Language Workbenches in general.



Even though modern IDEs with Language Servers come pretty close to what a syntax-based editor could do, I don't really like the architecture of having a server that needs to repeatedly parse the source code and then communicates ranges to the editor.

The real strength of a syntax-based editor is revealed when it can create editors for any language just by specifying the grammar.
Add the ability to define code transformations and you have a Language Workbench.

I have already read a bunch of research papers on the topic and tried JetBrains MPS, but didnt like it. Other syntax-based editors or Language Workbenches I only read about and didnt have hands-on experience so far.

**Reply by yairchu** [reply]
Hi Michael, nice meeting you :)

Similarly I've also collected frustrations and fantasies since at least when I saw Jonathan Edwards elaborate on the ills of text editing back in 2005.. I'm on a journey trying to create the "ideal programming language" with Lamdu :) Though not aiming to replace Rust/etc for the low-level stuff.

> I don't really like the architecture of having a server that needs to repeatedly parse the source code and then communicates ranges to the editor.

Can't agree more!

Breaking out of this scheme allows to have all IDE services available at all time and enables better type-error blame assignments according to which part of the code you changed, which imho really changes the game for trade-offs of dynamic vs static typing (elaborated in more detail here https://youtu.be/viF1bVTOO6k).

> create a shell-like command interpreter where you enter the commands via syntax-based editing

Cool! Do you have any demos of your project?

**Reply by michaelsippel** [reply]
Nice, lets keep in touch then.   I will take a closer look at lambdu soon, looks promising!

A very short demo of my project can be seen here: https://asciinema.org/a/580755
I'll release the source and more documentation about it in the next days.
